python/pyspark/sql/udf.py

```diff
@@ -157,7 +157,17 @@ def _create_judf(self):
         spark = SparkSession.builder.getOrCreate()
         sc = spark.sparkContext
 
-        wrapped_func = _wrap_function(sc, self.func, self.returnType)
+        func = fail_on_stopiteration(self.func)
+
+        # prevent inspect to fail
+        # e.g. inspect.getargspec(sum) raises
+        # TypeError: <built-in function sum> is not a Python function
+        try:
+            func._argspec = _get_argspec(self.func)
+        except TypeError:
```

Comment:
I would do:

```
    if evalType in (PythonEvalType.SQL_SCALAR_PANDAS_UDF,
                    PythonEvalType.SQL_GROUPED_MAP_PANDAS_UDF,
                    PythonEvalType.SQL_GROUPED_AGG_PANDAS_UDF):
        # the hack
```

to reduce the scope being affected by this logic more. Did this fix not working in your local?

this fix is rather a hack and we should clean up after merging this in ...

Reply:
the only way to do it there is to pass another argument to `UserDefinedFunction.__init__` with the wrapped function, because `UserDefinedFunction._wrapped` needs the original function (for docstring, etc.)

so I think what you suggest is actually messier (but ofc I will do it if you think it's the better way)

Reply:
Eh, don't we have `self.evalType` and I thought we could simply check it? I got that the current way is the "recommended" way to deal with divergence in Python but let's just explicitly scope here to make it easier to be fixed.

Reply:
Aha, you mean to do the hack in `UserDefinedFunction._create_judf`, and check the condition there? Sorry, I thought were talking about `_create_udf`. Wilco then

Reply:
Let's use all of Pandas UDFs:

https://github.com/apache/spark/blob/a9350d7095b79c8374fb4a06fd3f1a1a67615f6f/python/pyspark/sql/udf.py#L40-L67

Reply:
Also, let's leave a comment saying like this argspec is used for Pandas UDFs and the hack is to keep the original signature of given functions since there seem no way to copy it in Python 2.